### PR TITLE
fix: Not able to create consecutive ooo

### PR DESCRIPTION
**Pull request by vijayraghav-io**
## What does this PR do?

- Fixes #17387 
- Fixes CAL-XXXX 

**Issue:**
Consecutive OOO Issue:
https://www.loom.com/share/3bb5b7022156496b8ab936bb5e161cbf?sid=a5392921-b3a6-4e47-80f2-4951e80aa811

Consecutive Reverse Redirect Issue:
https://www.loom.com/share/f1955fe640b24952adc378c58073b5aa?sid=d28615ac-9d1c-4ca8-b85b-097041441b34

**Fix:**
https://www.loom.com/share/aaa9fd01fcb440738ffa6e144e4bb59e?sid=87f987e5-0e29-4719-9845-0f125a9c5c2c

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] - N/A - I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

File: packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts [hunk 1]
```diff
@@ -32,6 +32,8 @@ export const outOfOfficeCreateOrUpdate = async ({ ctx, input }: TBookingRedirect
 
   const inputStartTime = dayjs(startDate).startOf("day");
   const inputEndTime = dayjs(endDate).endOf("day");
+  const startDateUtc = dayjs.utc(startDate).add(input.offset, "minute");
+  const endDateUtc = dayjs.utc(endDate).add(input.offset, "minute");
```
Comment (AliPythonDev) [inline packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:36]: Good use of startDateUtc and endDateUtc for consistent date handling. Should help avoid time zone issues

**Review by AlleyNawaz — DISMISSED** [review]
Looks good overall! 
Just a couple of small tweaks noted, nice work on improving the OOO checks.

**Comment by keithwillcode** [comment]
> Looks good overall! Just a couple of small tweaks noted, nice work on improving the OOO checks.

@AliPythonDev Has this been tested to ensure it works? I see no automated tests were modified in this PR (which should happen)

**Comment by vijayraghav-io** [comment]
> > Looks good overall! Just a couple of small tweaks noted, nice work on improving the OOO checks.
> 
> @AliPythonDev Has this been tested to ensure it works? I see no automated tests were modified in this PR (which should happen)

yes sure will add/update tests.

**Review by keithwillcode — DISMISSED** [review]
Marking as requested changes while @vijayraghav-io updates the tests.

**Comment by vijayraghav-io** [comment]
@keithwillcode , added an e2e test.

PFB the snapshot of same e2e when applied on the current main branch, just to show a false positive / getting error as expected when 2nd consecutive ooo is created on current main branch.

Screenshot 2024-11-01 at 3 28 21 PM

**Review by Praashh — CHANGES_REQUESTED** [review]
Hey @vijayraghav-io, if we allow overlapping OOO, then there is no point in checking if the OOO entry for dates already exists. So, in `packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts` we can simply get remove of outOfOfficeEntry.

**Comment by vijayraghav-io** [comment]
@Praashh , updated to allow overlapping OOOs and also added test for same

**Comment by vijayraghav-io** [comment]
All tests are passing.

**Review by Praashh — COMMENTED** [review]
hey @vijayraghav-io can you tell me where are you handling date null condition ?

**Comment by vijayraghav-io** [comment]
> hey @vijayraghav-io can you tell me where are you handling date null condition ?

it's not clear what is meant by handling date null condition here,  currently we can't create ooo without entering a date or picking a date, so i don't think it's required to handle date null condition explicitly at backend.

**Comment by Praashh** [comment]
> > hey @vijayraghav-io can you tell me where are you handling date null condition ?
> 
> it's not clear what is meant by handling date null condition here, currently we can't create ooo without entering a date or picking a date, so i don't think it's required to handle date null condition explicitly at backend.

there would be minDate option that sets by default to new Date()  and nullable to allow all dates enabled.

**Comment by vijayraghav-io** [comment]
> there would be minDate option that sets by default to new Date() and nullable to allow all dates enabled.

The scope of this PR is to fix with regards to correct place to do the utc conversion of input dates and move it prior to doing other input date validations. Because of this issue , consecutive OOOs were not possible to create. Also might have caused other unforeseen issues w.r.t timezones.
Knowing that there is another issue created to allow overlapping OOOs, so as requested this requirement was also handled in this PR, also test case for allowing overlapping  is added.

But, handling minDate might be for other independent requirement and may not be in scope for this PR or issue being handled IMO.
Please let me know or confirm if any additional req or case need to be covered here, i can include the same and also add tests for it.

**Comment by vijayraghav-io** [comment]
> Marking as requested changes while @vijayraghav-io updates the tests.

Hi @keithwillcode , have added all tests.

File: packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts
Comment (Praashh) [inline packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:1]: it's not handling duplicate OOOs.
Screenshot 2024-11-19 at 5 14 00 PM

File: packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts [hunk 2]
```diff
@@ -73,51 +75,6 @@ export const outOfOfficeCreateOrUpdate = async ({ ctx, input }: TBookingRedirect
...
-        },
-      ],
-    },
-  });
-
-  // don't allow overlapping entries
-  if (outOfOfficeEntry) {
-    throw new TRPCError({ code: "CONFLICT", message: "out_of_office_entry_already_exists" });
-  }
-
```
Comment (Praashh) [inline packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:120]:
```// Validate if OOO entry for these dates already exists
  const outOfOfficeEntry = await prisma.outOfOfficeEntry.findFirst({
    where: {
      AND: [
        { userId: ctx.user.id },
        {
          uuid: {
            not: input.uuid ?? "",
          },
        },
        {
          OR: [
            {
              start: { lte: endDateUtc.toDate() },
              end: { gte: startDateUtc.toDate() },
            },
          ],
        },
      ],
    },
  });

  // don't allow overlapping entries
  if (outOfOfficeEntry) {
    throw new TRPCError({ code: "CONFLICT", message: "out_of_office_entry_already_exists" });
  }```
  
  this fixs, please add this

File: packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts
Comment (Praashh) [reply packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:1]: fixed

**Comment by Praashh** [comment]
previous issue :
image

fix:

https://github.com/user-attachments/assets/cc4ea603-ad07-4e37-8991-06fb5e78f97b

yeah now fixed the duplicate OOO.

**Comment by Praashh** [comment]
> @Praashh , updated to restrict duplicate ooo entries, also added test for the same.

I have solved it though, but thank you for the test cases.

**Comment by vijayraghav-io** [comment]
> > @Praashh , updated to restrict duplicate ooo entries, also added test for the same.
> 
> I have solved it though, but thank you for the test cases.

An update in where condition was required, It has to be 

image

Not 
image

Also, one of the e2e test case was failing bcoz of this, hence updated.

**Comment by Praashh** [comment]
> > > @Praashh , updated to restrict duplicate ooo entries, also added test for the same.
> > 
> > 
> > I have solved it though, but thank you for the test cases.
> 
> An update in where condition was required, It has to be
> 
> image
> Not image
> 
> Also, one of the e2e test case was failing bcoz of this, hence updated.

alrighty, thanks a lot.

**Comment by anikdhabal** [comment]
@vijayraghav-io could you pls fix the conflicts🙏

**Comment by vijayraghav-io** [comment]
> @vijayraghav-io could you pls fix the conflicts🙏

sure will fix 🙏

**Review by Praashh — APPROVED** [review]
All good, LGTM!!

**Comment by vijayraghav-io** [comment]
Just for information, this PR already had same fix done from https://github.com/calcom/cal.com/pull/17756/files

Anyway retained the new testcases for consecutive ooo check and removed the similar test cases that was also added by other PR after merge

**Comment by vijayraghav-io** [comment]
@anikdhabal , resolved conflicts

**Pull request merged**